p2p/chain.py

```diff
@@ -209,29 +215,36 @@ def register_peer(self, peer: BasePeer) -> None:
             if now - req_time > self._reply_timeout
         ]
         while timed_out:
-            self.logger.warn("Re-requesting %d timed out block parts...", len(timed_out))
+            self.logger.warn(
+                "Re-requesting %d timed out %s out of %d pending ones",
+                len(timed_out), part_name, len(pending))
             await request_func(timed_out[:batch_size])
             timed_out = timed_out[batch_size:]
 
     async def body_downloader(self) -> None:
         await self._downloader(
             self._body_requests,
-            self._should_skip_body_download,
+            self._skip_empty_bodies,
             self.request_bodies,
             self._pending_bodies,
-            eth.MAX_BODIES_FETCH)
+            eth.MAX_BODIES_FETCH,
+            'bodies')
 
     async def receipt_downloader(self) -> None:
         await self._downloader(
             self._receipt_requests,
-            self._should_skip_receipts_download,
+            self._skip_empty_and_duplicated_receipts,
             self.request_receipts,
             self._pending_receipts,
-            eth.MAX_RECEIPTS_FETCH)
+            eth.MAX_RECEIPTS_FETCH,
+            'receipts')
 
-    def _should_skip_body_download(self, header: BlockHeader) -> bool:
-        return (header.transaction_root == self.chaindb.empty_root_hash and
-                header.uncles_hash == EMPTY_UNCLE_HASH)
+    def _skip_empty_bodies(self, headers: List[BlockHeader]) -> List[BlockHeader]:
+        return [
+            header
+            for header in headers
+            if (header.transaction_root != self.chaindb.empty_root_hash or
```

Comment:
Probably improved readability if you moved this out of a comprehension and instead used the `to_tuple` decorator to do this in a proper for loop.  It would allow this logic to be broken across multiple lines in an easier to read manner.

Reply:
Makes sense. Doing that on #492 though as that's where it was introduced